### ISSUE TITLE
Configs: include `return-await` in `strict-type-checked` preset

### ISSUE THREAD

**Issue by aseemk** [typescript-eslint/typescript-eslint#8667]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22preset+config+change%22+) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Description

I recently debugged a difficult bug in my code where I was try-catching an async call that I was returning right away.

```ts
async function asyncFoo() {
  try {
    return asyncBar() // no await
  } catch (err) {
    // so my error handling code here wasn't getting hit
  }
}
```

I was surprised when I finally figured this out, because I love and rely on e.g. [`no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises/) to catch missing `await`s, but it hadn't caught this one.

I found that [`return-await`](https://typescript-eslint.io/rules/return-await/) exists to help here, but was then also surprised that it wasn't included in the [`recommended-type-checked`](https://typescript-eslint.io/users/configs#recommended-type-checked) preset that we use.

I thought the whole purpose of that preset was to at least help ensure _correctness_ (separate from style) — and this was definitely a correctness issue.

I've now enabled `return-await` in my codebase, and am feeling great. But can I suggest adding this to `recommended-type-checked` to prevent other developers from hitting easily preventable bugs like this?

Thank you for your consideration and for a great tool!

### Impacted Configurations

[`recommended-type-checked`](https://typescript-eslint.io/users/configs#recommended-type-checked)

**Comment by JoshuaKGoldberg** [comment]
Heh, I personally prefer this, but have memories of people getting ✨ _very upset_ ✨ over it. Maybe we should put it in `strict` to start?

**Comment by Josh-Cena** [comment]
Why would people get upset about `return-await`? Returning an unawaited promise inside a try catch seems to invite for troubles. I also don't know why people ever want to return unawaited promises, except for very simple wrappers—fun fact, awaiting the promise does not make anything slower; it may actually make code faster than if you return the promise directly!

**Comment by JoshuaKGoldberg** [comment]
I don't have many primary resources that I can find in 5 minutes of searching 😞 but vaguely:

* The venerable https://jakearchibald.com/2017/await-vs-return-vs-return-await explicitly says _"Outside of try/catch blocks, `return await` is redundant_ - and that article still has good SEO
* When https://eslint.org/docs/latest/rules/no-return-await was written, IIRC it was taken to heart by many
* Scattered old tweets from searching for `return await javascript` show folks occasionally suggesting a preference for less code (`return a` over `return await a`) and/or not wanting to pollute the call stack

**Comment by kirkwaiblinger** [comment]
I'm big +1 to this, but, I am also concerned that the current rule is too opinionated, and would cause a lot of unnecessary flagging in existing codebases. I'd propose we have an unopinionated option that only flags on the parts of the rule that impact correctness, if we are to add the rule to recommended; see #9030

**Comment by bradzacher** [comment]
> and would cause a lot of unnecessary flagging in existing codebases

The rule has an autofixer for any cases that aren't `any`/`unknown` - so why is extra **_initial_** flagging bad?
It brings the codebase into a consistent style immediately. Anecdotally I've found people generally don't write `return await` so I don't think it would flag much.

**Comment by JoshuaKGoldberg** [comment]
> people generally don't write `return await`

...but they should! If you `return myPromise` instead of `return await myPromise` the call stack doesn't include the intermediary step. 

I'm +1 to this provided the caveats Kirk mentioned are accounted for.

**Comment by kirkwaiblinger** [comment]
> so why is extra **_initial_** flagging bad? 

Mostly, it's just that I would be scared we'd not get the rule into a recommended preset at all if we reawaken the opinion-charged conversations referenced in https://github.com/typescript-eslint/typescript-eslint/issues/8667#issuecomment-2020178418. The thought is that, strategically speaking, we could sidestep all that and still get the most important benefits (IMO) of return-await.

> It brings the codebase into a consistent style immediately. Anecdotally I've found people generally don't write `return await` so I don't think it would flag much.

I completely agree that full consistency (whether `in-try-catch` or, my preference, `always`) should be the correct end goal for any codebases

**Comment by kirkwaiblinger** [comment]
If backlash that it's too opinionated where it doesn't affect the outcome of execution is not a concern, then everything I've said is moot!

**Comment by JoshuaKGoldberg** [comment]
@kirkwaiblinger I defer to your judgment on which preset to add this to 🙂 _(not speaking for other maintainers)_

**Comment by bradzacher** [comment]
> but they should! If you return myPromise instead of return await myPromise the call stack doesn't include the intermediary step.

This isn't true any more AFAIK.
Chrome et al actually do a pretty good job of correctly tracking the context without `await`:

```js
async function throws() {
  throw new Error('one');
}

function noAwaitSync() {
  return throws();
}
async function noAwaitAsync() {
  return throws();
}
async function awaitAsync() {
  return await throws();
}
```

| Chrome | FF |
| --- | --- |
| ![console screenshot 1](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/df2b82fe-dc30-4098-aed1-9676a355cd38) | ![console screenshot 2](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/a6bd42f0-7df2-488e-9029-c3df9fd04372) |

(note FF eagerly evaluates AND LOGS THE ERROR - hence the double erroring)

**Comment by kirkwaiblinger** [comment]
TL;DR let's turn on `return-await` in `recommended-type-checked` contingent on blocking issues being resolved.

Kirk's order of preference for which option to use
1. correctness only option (https://github.com/typescript-eslint/typescript-eslint/issues/9030) in recommended-type-checked, "always" in strict-type-checked
1. (close second) "always" across the board
1. (distant third) "in-try-catch" across the board
***

Having given this some thought, I here's where I'm at.

## Config Choice

This rule has the potential to prevent very subtle bugs around control flow that are not really just a matter of `strict`-config-level pedantry, so I think it's justifiable to be enabled in `recommended-type-checked`. 

## Blockers

This rule also has the potential to _cause_ very subtle bugs in control flow if it has bugs in its implementation (particularly in its autofixer). Therefore, I think that the following are _blockers_ for introducing it into _any_ config:
1. https://github.com/typescript-eslint/typescript-eslint/issues/8661
4. https://github.com/typescript-eslint/typescript-eslint/issues/8907
5. https://github.com/typescript-eslint/typescript-eslint/issues/7889

The first two are well on their way in https://github.com/typescript-eslint/typescript-eslint/pull/9031, and the third can be actioned as soon as that PR goes through (see https://github.com/typescript-eslint/typescript-eslint/pull/9044), so I think those will all be resolved relatively quickly.

## Options Choice

`in-try-catch`, `always`, and the proposed "correctness-only" option (https://github.com/typescript-eslint/typescript-eslint/issues/9030) all agree on the subset of the rule reports that impact correctness, i.e. error-handling contexts.

For reports _outside_ error-handling contexts, `in-try-catch` and `always` enforce a conflicting stylistic preference that is 100% autofixable (which will still be the case after the changes in the aforementioned PRs go through). 

I personally would prefer choosing neither, instead creating a "correctness-only" option (https://github.com/typescript-eslint/typescript-eslint/issues/9030) for the recommended config, but there doesn't seem to be much enthusiasm for that from anyone else 😆. 

Due to the autofixability, I do buy the argument in https://github.com/typescript-eslint/typescript-eslint/issues/8667#issuecomment-2083520373, and I think it's acceptable to pick either `in-try-catch` or `always` in the recommend config.  Users can easily resolve it to their preference by just changing the option. 

In a vacuum, I think that `always` is a _much, much_ better choice, between "always" and "in-try-catch". 
1. (important) If there are any more bugs in the rule about what is and isn't an error-handling context, whether a correctness bug, like https://github.com/typescript-eslint/typescript-eslint/issues/8907, or simply a gap in support, like https://github.com/typescript-eslint/typescript-eslint/issues/7889, _`always` will still get it right, rather than enforce bug vectors in people's code_. 
1. (not important) I also just happen to like the explicitness better as a reader

But, `in-try-catch` is the current default for the rule... and it may align better with more users' expectations. In any case, if we put `always` in the config, we should also probably change the rule default.

**Comment by Josh-Cena** [comment]
> Do you have references?

Sorry I lost track of this. See https://github.com/tc39/proposal-faster-promise-adoption

**Comment by kirkwaiblinger** [comment]
I think this rule is good to go into recommended-type-checked. 

Side conversations ongoing regarding the rule option, wrapping up shortly, though.

There are IMO three open bugs that are blockers, though: 
1. https://github.com/typescript-eslint/typescript-eslint/issues/8661
1. https://github.com/typescript-eslint/typescript-eslint/issues/8907
1. https://github.com/typescript-eslint/typescript-eslint/issues/7889

The first two are well on their way in https://github.com/typescript-eslint/typescript-eslint/pull/9031, and the third can be actioned as soon as that PR goes through (see https://github.com/typescript-eslint/typescript-eslint/pull/9044), so I think those will all be resolved relatively quickly.

**Comment by kirkwaiblinger** [comment]
@bradzacher @Josh-Cena @JoshuaKGoldberg

ok! So, summarizing some side, conversations, I _think_ the least common denominator that we've landed on now is that we can do strict-type-checked, but that we _do_ need to resurrect https://github.com/typescript-eslint/typescript-eslint/issues/9030 in order to do that. 

we could put return-await in stylistic-type-checked, too, if we want, with either `always` or `in-try-catch`. If someone wants to propose to make that happen, please spin that off into a separate issue 🙂
